### PR TITLE
Make type promotion match PyTorch

### PR DESCRIPTION
**Pull request by ailzhang**
* `PromoteSecond` is not deleted because Pytorch passes wrapped Python number as highest-precision number, but that precision is not considered in type promotion. In XLA `Promote` we have no way to know whether a shape is from a wrapped number or not. 
This can be removed in a followup PR after I confirmed the PT plan for scalar, currently we have `at::Scalar` as well as `at::Tensor` with `is_wrapped_number=True`, which is weird..
* I copied the python test suite for type promotion over just to prevent breakage(types are very easily broken with changes). We can remove this once the PT test is moved to generic test.

### CODE REVIEW

File: test/test_operations.py [hunk 1]
```diff
@@ -1298,6 +1298,303 @@ def test_git_revisons(self):
     self.assertTrue(revs['xla'])
     self.assertTrue('torch' in revs)
 
+class TestTypePromotion(TestCase):
```
Comment (dlibenzi) [inline test/test_operations.py]: This is 4 spaces indented ... I am afraid passing it with `google-pyformat` might make a mess.
Can you try to bring it to 2 spaces manually?

File: torch_xla/csrc/aten_xla_bridge.cpp [hunk 2]
```diff
@@ -64,6 +65,16 @@ XLATensor GetXlaTensor(const at::Tensor& tensor) {
   return *xtensor;
 }
 
+std::tuple<XLATensor, XLATensor> GetPromotedXlaTensorsForBinaryOp(
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_bridge.cpp]: This belongs more to `tensor_ops.*` than here.

File: torch_xla/csrc/aten_xla_bridge.h [hunk 3]
```diff
@@ -23,6 +23,10 @@ XLATensor GetXlaTensor(const at::Tensor& tensor);
 std::vector<XLATensor> GetXlaTensors(
     tensorflow::gtl::ArraySlice<const at::Tensor> tensors);
 
+// Same as above, with type promotion
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_bridge.h]: Ditto ... `tensor_ops.*` ... and this is most definitely not like the above 😉

File: torch_xla/csrc/aten_xla_type.cpp [hunk 4]
```diff
@@ -2009,8 +2006,9 @@ at::Tensor AtenXlaType::mul(const at::Tensor& self, at::Scalar other) {
 
 at::Tensor& AtenXlaType::mul_(at::Tensor& self, const at::Tensor& other) {
   XLATensor self_tensor = bridge::GetXlaTensor(self);
-  XLATensor::mul_(self_tensor,
-                  bridge::GetOrCreateXlaTensor(other, self_tensor.GetDevice()));
+  XLATensor other_tensor =
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: Why this change?

File: torch_xla/csrc/tensor.h [hunk 5]
```diff
@@ -908,6 +908,9 @@ class XLATensor {
   static XLATensor where(const XLATensor& condition, const XLATensor& input,
                          const XLATensor& other);
 
+  // Set logical_element_type which is visible to upstream PyTorch.
+  void SetScalarType(c10::optional<at::ScalarType> logical_element_type);
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor.h]: This should go up close to the `dtype()` function.

File: torch_xla/csrc/aten_xla_bridge.cpp (hunk 2 above)
Comment (ailzhang) [reply torch_xla/csrc/aten_xla_bridge.cpp]: Hmmmm I think it fit here better since it's taking `at::Tensor` as inputs and return `XLATensor` which is mostly what the bridge does? ;)

File: torch_xla/csrc/aten_xla_type.cpp (hunk 4 above)
Comment (ailzhang) [reply torch_xla/csrc/aten_xla_type.cpp]: Yea I forgot to revert this change. Nice catch!

**Comment by ailzhang** [comment]
> I am a bit confused though.
> This PR was supposed to be cleaning up all the `PromoteSecond()` business, which is still there ...

I tried :P Everything passed if we remove `PromoteSecond` except for `float_tensor.add_(0.5)` case. 
So we have 2 type systems: pytorch scalartype and XLA primitives. Type promotion is not applied to inplace ops, so we use `Promote` in xla, this will return a double tensor which is not expected. 
To fix it, there're two ways:
1. Let XLA type promotion be aware of wrapped numbers and exclude it in type promotion.
2. Fix on PT side that 0.5 should be match the first dtype instead of always using highest precision for wrapped numbers. 
I personally think option 2 is the right way to go as it's easy and straightforward. And option 1 is definitely not good as it requires XLA to inspect not only shapes but the source of tensors (wrapped number and scalar tensor have exactly the same shape, but behave differently in type promotion). 
So in the end I brought back `PromoteSecond` temporarily, and I'm planning to remove it once I confirm option 2 is okay with upstream. :D

File: torch_xla/csrc/aten_xla_bridge.cpp (hunk 2 above)
Comment (dlibenzi) [reply torch_xla/csrc/aten_xla_bridge.cpp]: You are right ... tensor_ops are ops among XLA tensors.
But they do not belong here as well. The bridge should not know about binary ops and type promotion.
Given that the only caller is aten_xla_type.cpp put this in the anon namespace there.

**Comment by dlibenzi** [comment]
> > I am a bit confused though.
> > This PR was supposed to be cleaning up all the `PromoteSecond()` business, which is still there ...
> 
> I tried :P Everything passed if we remove `PromoteSecond` except for `float_tensor.add_(0.5)` case.
> So we have 2 type systems: pytorch scalartype and XLA primitives. Type promotion is not applied to inplace ops, so we use `Promote` in xla, this will return a double tensor which is not expected.
> To fix it, there're two ways:
> 
> 1. Let XLA type promotion be aware of wrapped numbers and exclude it in type promotion.
> 2. Fix on PT side that 0.5 should be match the first dtype instead of always using highest precision for wrapped numbers.
>    I personally think option 2 is the right way to go as it's easy and straightforward. And option 1 is definitely not good as it requires XLA to inspect not only shapes but the source of tensors (wrapped number and scalar tensor have exactly the same shape, but behave differently in type promotion).
> 
> So in the end I brought back `PromoteSecond` temporarily, and I'm planning to remove it once I confirm option 2 is okay with upstream. :D

Option 2 works for me but I wonder why that would fail.
AFAICT we have:
```C++
void XLATensor::add_(XLATensor& input, at::Scalar other, at::Scalar alpha);
```
Then we call:
```C++
GetIrValueForScalar(other, input.shape(), input.GetDevice());
```
Passing `input` shape.
That is going to create an IR Value with the same XLA type of `input`.
So the following:
```C++
input.SetIrValue(input.GetIrValue() + other_constant * alpha_constant); 
```
Should not make any promotion and just keep `input` type (`input`, `other_constant` and `alpha_constant` should all have the same type).

**Comment by ailzhang** [comment]
@dlibenzi The sad fact is that when you do `a + 0.3` PT actually dispatch to `add(Tensor, Tensor)` directly that doesn't go through `add(Tensor, Scalar)`.

**Comment by dlibenzi** [comment]
> @dlibenzi The sad fact is that when you do `a + 0.3` PT actually dispatch to `add(Tensor, Tensor)` directly that doesn't go through `add(Tensor, Scalar)`.

Oh, you mentioned `float_tensor.add_(0.5)`, this is why I made the example above.

File: torch_xla/csrc/aten_xla_bridge.cpp [hunk 6]
```diff
@@ -9,6 +9,7 @@
 #include "tensorflow/compiler/xla/xla_client/debug_macros.h"
 #include "torch_xla/csrc/device.h"
 #include "torch_xla/csrc/tensor_impl.h"
+#include "torch_xla/csrc/tensor_util.h"
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_bridge.cpp]: You left this around :wink:
